app/uk/gov/hmrc/excisemovementcontrolsystemapi/scheduling/PollingNewMessageWithWorkItemJob.scala

```diff
@@ -0,0 +1,156 @@
+            case (false, _) if wi.failureCount < retryAttempt => workItemRepository.markAs(wi.id, ProcessingStatus.Failed)
+            case _ => workItemRepository.markAs(wi.id, ProcessingStatus.PermanentlyFailed)
+          }
+        }.flatMap(_ => process(failedBefore, availableBefore, retryAttempt))
+      }
+      .recoverWith {
+        case NonFatal(e) =>
+          logger.error("[PollingNewMessageWithWorkItemJob] - Failed to collect and process ann the movement", e)
+          Future.failed(RunningOfJobFailed(name, e))
+      }
+  }
+
+
+  private def getNewMessages(exciseNumber: String): Future[(Boolean, NewMessageResult)] = {
+    newMessageService.getNewMessagesAndAcknowledge(exciseNumber)
+      .flatMap(message =>
+        message match {
+          case Some(response) if response.message.nonEmpty =>
+            saveToDB(exciseNumber, response).map(result => (result, MessageReceived))
+          case _ => Future.successful((true, NoMessageFound))
+        })
+      .recover {
+        case NonFatal(e) =>
+          logger.error(s"[PollingNewMessageWithWorkItemJob] - Could not get messages for ern: ${exciseNumber} with message: ${e.getMessage}. Will retry later", e)
+          (false, MessageError)
+      }
+  }
+
+  private def saveToDB(
+    exciseNumber: String,
+    newMessageResponse: EISConsumptionResponse
+  ): Future[Boolean] = {
+
+    val messages = messageParser.extractMessages(newMessageResponse.message)
+
+    //! process IE801 or IE704 first if any.We are Processing message in sequence (not in parallel)
+    // If we want to process them in parallel we need to put a lock on mongo when reading
+    // and writing. As we could read just before  or while writing and the next time we write we
+    // may do not write the up to date info. At the moment we have a mongo lock on the Job.
+    saveAcceptedOrRefusalMessage(exciseNumber, messages)
+    saveRestOfMessages(exciseNumber, messages)
+  }
+
+  private def save(message: IEMessage, exciseNumber: String)(implicit ec: ExecutionContext): Future[Boolean] = {
+    movementService.updateMovement( message, exciseNumber)
+      .flatMap {
+        case true =>
+          successful(true)
+        case _ =>
+          logger.warn(s"[PollingNewMessageWithWorkItemJob] - Could not update movement for ern: $exciseNumber")
+          successful(false)
+      }
+  }
+
+  private def saveAcceptedOrRefusalMessage(exciseNumber: String, messages: Seq[IEMessage]): Future[Boolean] = {
+    messages.filter(isAcceptedOrRefusalMessage(_))
+      .foldLeft(successful(true)) { case (acc, x) =>
+        acc.flatMap {
+          _ => save(x, exciseNumber)
+        }
+      }
+  }
+
+  private def saveRestOfMessages(exciseNumber: String, messages: Seq[IEMessage]): Future[Boolean] = {
+    messages
+      .filterNot(isAcceptedOrRefusalMessage(_))
+      .foldLeft(successful(true)) { case (acc, x) =>
+        acc.flatMap {
+          _ => save(x, exciseNumber)
+        }
+      }
+  }
```

Comment:
Both these methods are doing the same thing effectively but the only difference is how the list is filtered. This could be one method that takes the filter lambda as a parameter and save the messages based on the call to the filter.

Reply:
I would prefer to stay two method as the function describe what they are doing more explicitly. I could refactor by both taking a common function, but living the functions (for readiness)